### PR TITLE
[ncl] Upgrade `@react-navigation` to v6

### PR DESCRIPTION
**Pull request by bbarthec**
# Why

`ncl` project fails on `main` branch with following error:
```
TypeError: undefined is not a constructor (evaluating 'new _reactNativeReanimated.default.Value(0)')
at /Users/bbarthec/Work/expo/react-native-lab/react-native/Libraries/Core/ExceptionsManager.js:95:4 in reportException
at /Users/bbarthec/Work/expo/react-native-lab/react-native/Libraries/Core/ExceptionsManager.js:141:19 in handleException
at /Users/bbarthec/Work/expo/react-native-lab/react-native/Libraries/Core/setUpErrorHandling.js:24:6 in handleError
at /Users/bbarthec/Work/expo/node_modules/@react-native/polyfills/error-guard.js:49:36 in ErrorUtils.reportFatalError
at /Users/bbarthec/Work/expo/node_modules/metro-runtime/src/polyfills/require.js:203:6 in guardedLoadModule
at http://192.168.1.45:19001/__generated__/AppEntry.bundle?platform=android&dev=true&hot=false&strict=false&minify=false:506830:3 in global code

Invariant Violation: "main" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called.
at /Users/bbarthec/Work/expo/react-native-lab/react-native/Libraries/Core/ExceptionsManager.js:95:4 in reportException
at /Users/bbarthec/Work/expo/react-native-lab/react-native/Libraries/Core/ExceptionsManager.js:141:19 in handleException
at /Users/bbarthec/Work/expo/react-native-lab/react-native/Libraries/Core/setUpErrorHandling.js:24:6 in handleError
at /Users/bbarthec/Work/expo/node_modules/@react-native/polyfills/error-guard.js:49:36 in ErrorUtils.reportFatalError
```
This error is coming from usage of `reanimated@v1` by `@react-navigation/drawer` library.

Partially resolves [ENG-4150](https://linear.app/expo/issue/ENG-4150/native-component-list)

# How

1. Removed unused v4 `react-navigation` dependencies
2. Bumped every `@react-navigation/*` library to the newest one
3. Adapted codebase a bit to work and look fine following this upgrade guide: https://reactnavigation.org/docs/upgrading-from-5.x/
    - this step is not performed very accurately, as I'm not an expert in terms of `@react-navigation`, especially `linking` part might be broken somewhere 👀 

# Test Plan

`ncl` no longer crashes in runtime

### CODE REVIEW

**Review by lukmccall — APPROVED** [review]

**Comment by brentvatne** [comment]
related task: https://linear.app/expo/issue/ENG-4148/upgrade-react-navigation-across-expoexpo-monorepo

**Comment by bbarthec** [comment]
I'll merging this PR now, because we need it to unblock manual testing using `ncl`. Anyone interested: feel free to do after-merge review or/and propose another PR fixing things 😉 
Disclaimer: I see half of CI jobs are failing, but the reasons for failures are not-related to this PR. CI need proper maintenance right now 😭

**Pull request merged**
